File: app/src/main/java/com/song/deviceinfo/ui/applist/AppListFragment.java
package com.song.deviceinfo.ui.applist;

import com.song.deviceinfo.model.beans.ApplicationBean;
import com.song.deviceinfo.ui.base.BaseAdapter;
import com.song.deviceinfo.ui.base.BaseFragment;
import com.song.deviceinfo.ui.base.BaseViewModel;
import com.song.deviceinfo.utils.LogUtils;
import com.song.deviceinfo.utils.ThreadPoolUtils;

import org.json.JSONArray;

import java.util.List;

import androidx.lifecycle.ViewModelProviders;

/**
 * Created by chensongsong on 2020/6/3.
 */
public class AppListFragment extends BaseFragment<ApplicationBean> {

    @Override
    protected BaseAdapter crateAdapter() {
        return new AppListAdapter(getContext());
    }

    @Override
    protected BaseViewModel crateViewModel() {
        return ViewModelProviders.of(this).get(AppListViewModel.class);
    }

    @Override
    protected void refreshData() {
        ThreadPoolUtils.getInstance().execute(() -> {
            final List<ApplicationBean> list = getAppListInfo();
            JSONArray jsonArray = new JSONArray();
            for (ApplicationBean bean : list) {
                try {
                    jsonArray.put(bean.toJSON());
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            LogUtils.printLongString(jsonArray.toString());
            mainHandler.post(() -> {
                viewModel.setValue(list);
            });
        });
    }

    private List<ApplicationBean> getAppListInfo() {
        return ((AppListViewModel) viewModel).getAppListInfo(getContext());
    }

}

File: app/src/main/java/com/song/deviceinfo/ui/applist/AppListFragment.kt
package com.song.deviceinfo.ui.applist

import androidx.lifecycle.ViewModelProviders
import com.song.deviceinfo.model.beans.ApplicationBean
import com.song.deviceinfo.ui.base.BaseAdapter
import com.song.deviceinfo.ui.base.BaseFragment
import com.song.deviceinfo.ui.base.BaseViewModel
import com.song.deviceinfo.utils.LogUtils.printLongString
import org.json.JSONArray

/**
 * Created by chensongsong on 2020/6/3.
 */
class AppListFragment : BaseFragment<ApplicationBean>() {
    override fun createAdapter(): BaseAdapter<ApplicationBean, *> {
        return AppListAdapter(requireContext())
    }
    
    override fun createViewModel(): BaseViewModel<ApplicationBean> {
        return ViewModelProviders.of(this).get(AppListViewModel::class.java)
    }
    
    override fun refreshData() {
        launchOnIO {
            val list = appListInfo
            val jsonArray = JSONArray()
            for (bean in list) {
                try {
                    jsonArray.put(bean.toJSON())
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            printLongString(jsonArray.toString())
            launchOnMain { viewModel?.setItems(list) }
        }
    }
    
    private val appListInfo: List<ApplicationBean>
        get() = (viewModel as AppListViewModel).getAppListInfo(requireContext())
}
